src/main/java/org/apache/sling/scripting/core/impl/ServiceCache.java

```diff
@@ -63,70 +70,151 @@ public void dispose() {
      * @return The service or <code>null</code>
      */
     @SuppressWarnings("unchecked")
+    @Nullable
     public <ServiceType> ServiceType getService(Class<ServiceType> type) {
-        final String key = type.getName();
-        Reference reference = this.cache.get(key);
-        if (reference == null) {
-
-            // get the service
-            ServiceReference ref = this.bundleContext.getServiceReference(key);
-            if (ref != null) {
-                final Object service = this.bundleContext.getService(ref);
-                if (service != null) {
-                    reference = new Reference();
-                    reference.service = service;
-                    reference.reference = ref;
-                } else {
-                    ref = null;
-                }
-            }
-
-            // assume missing service
-            if (reference == null) {
-                reference = NULL_REFERENCE;
+        SortedSet<Reference> references = getCachedReferences(type);
+        for (Reference reference : references) {
+            ServiceType service = (ServiceType) reference.getService();
+            if (service != null) {
+                return service;
             }
+        }
+        return null;
+    }
 
-            // check to see whether another thread has not done the same thing
-            synchronized (this) {
-                Reference existing = this.cache.get(key);
-                if (existing == null) {
-                    this.cache.put(key, reference);
-                    ref = null;
-                } else {
-                    reference = existing;
+    @SuppressWarnings("unchecked")
+    @Nullable
+    public <ServiceType> ServiceType[] getServices(Class<ServiceType> type, String filter) {
+        List<ServiceType> result = new ArrayList<>();
+        try {
+            SortedSet<Reference> cachedReferences = getCachedReferences(type);
+            final Collection<ServiceReference<ServiceType>> filteredReferences = this.bundleContext.getServiceReferences(type, filter);
+            if (!filteredReferences.isEmpty()) {
+                List<ServiceReference<ServiceType>> localFilteredReferences = new ArrayList<>(filteredReferences);
```

Comment:
I guess the extra list isn't really needed, right? We could just use Arrays.sort with reverse order for the sorting and the for loop already works with arrays (not that it matters much).

Reply:
We do need this list, as we need to sort the `ServiceReferences` according to their ranking, in order to be able to return the array with the `ServiceTypes`, the latter being objects that don't implement `Comparable`.